### ISSUE TITLE
The different results from same MLP.

### ISSUE THREAD

**Issue by mt-cly** [pytorch/pytorch#109371]
### 🐛 Describe the bug

Hi, I met a bug. There is two-layer MLP, I tried to print twice layer(x) in the forward function, unreasonablelly, the outputs of the second layer were different.  

System: Ubuntu18.04
Python: 3.10.11
Torch: 1.13.1+cu116 installed with the command:
```python
pip install torch==1.13.1+cu116 torchvision==0.14.1+cu116 torchaudio==0.13.1 --extra-index-url https://download.pytorch.org/whl/cu116'
```
The code for reproducing:

`
from torch import nn
from torch.nn import functional as F
import torch

class MLP(nn.Module):
    def __init__(
        self,
        input_dim: int,
        hidden_dim: int,
        output_dim: int,
        num_layers: int,
        sigmoid_output: bool = False,
    ) -> None:
        super().__init__()
        self.num_layers = num_layers
        h = [hidden_dim] * (num_layers - 1)
        self.layers = nn.ModuleList(
            nn.Linear(n, k) for n, k in zip([input_dim] + h, h + [output_dim])
        )
        self.sigmoid_output = sigmoid_output

    def forward(self, x):
        for i, layer in enumerate(self.layers):
            print(layer(x)[0,0])
            print(layer(x)[0,0])
            print(sum(layer.weight[0]*x[0])+layer.bias[0])
            x = F.relu(layer(x)) if i < self.num_layers - 1 else layer(x)
            continue
        if self.sigmoid_output:
            x = F.sigmoid(x)
        return x

if __name__ == '__main__':
    mlp = MLP(256,256,32, 2).cuda()
    x = torch.ones(1,256).cuda()
    results = mlp(x)
`

The log is as follows: 

```
tensor(0.7011, device='cuda:0', grad_fn=<SelectBackward0>)
tensor(0.7011, device='cuda:0', grad_fn=<SelectBackward0>)
tensor(0.7011, device='cuda:0', grad_fn=<AddBackward0>)
tensor(0.8312, device='cuda:0', grad_fn=<SelectBackward0>)
tensor(0.9613, device='cuda:0', grad_fn=<SelectBackward0>)
tensor(0.1888, device='cuda:0', grad_fn=<AddBackward0>)
```

When I uninstall this 1.13.1 pytorch vision and install another one like 1.12.0, this bug is eliminated. 

### Versions

Collecting environment information...
PyTorch version: 1.13.1+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.10.11 (main, Apr 20 2023, 19:02:41) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-150-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 11.1.74
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA RTX A6000
GPU 1: NVIDIA RTX A6000
GPU 2: NVIDIA RTX A6000
GPU 3: NVIDIA RTX A6000
GPU 4: NVIDIA RTX A6000
GPU 5: NVIDIA RTX A6000
GPU 6: NVIDIA RTX A6000
GPU 7: NVIDIA RTX A6000

Nvidia driver version: 470.182.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              144
On-line CPU(s) list: 0-143
Thread(s) per core:  2
Core(s) per socket:  36
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               106
Model name:          Intel(R) Xeon(R) Platinum 8373C CPU @ 2.60GHz
Stepping:            6
CPU MHz:             890.753
CPU max MHz:         3500.0000
CPU min MHz:         800.0000
BogoMIPS:            5200.00
Virtualization:      VT-x
L1d cache:           48K
L1i cache:           32K
L2 cache:            1280K
L3 cache:            55296K
NUMA node0 CPU(s):   0-35,72-107
NUMA node1 CPU(s):   36-71,108-143
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==1.13.1+cu116
[pip3] torchaudio==0.13.1+cu116
[pip3] torchvision==0.14.1+cu116
[conda] blas                      1.0                         mkl  
[conda] cpuonly                   2.0                           0    pytorch
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0           py310h7f8727e_0  
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0  
[conda] mkl_random                1.2.2           py310h00e6091_0  
[conda] numpy                     1.24.3          py310hd5efca6_0  
[conda] numpy-base                1.24.3          py310h8e6c178_0  
[conda] pytorch-mutex             1.0                         cpu    pytorch
[conda] torch                     1.13.1+cu116             pypi_0    pypi
[conda] torchaudio                0.13.1+cu116             pypi_0    pypi
[conda] torchvision               0.14.1+cu116             pypi_0    pypi

**Comment by mikaylagawarecki** [comment]
Does this reproduce for you on 2.0.1/the latest main? I was unable to reproduce this on both, could you upgrade your PyTorch to the latest version and retry please?

**Comment by mt-cly** [comment]
Hi, thanks for your reply.
I have checked with two versions (the latest 2.0.1 & the previous 1.12.0), and things go well without bug. It seems that this bug occurs only in pytorch 1.13.1

**Comment by albanD** [comment]
Thanks for checking!
Will close as fixed then!
